src/DynamoCore/Logging/DynamoAnalyticsClient.cs

```diff
@@ -167,102 +157,228 @@ public void Start()
                 RegisterADPTracker(service);
 
                 //If not ReportingAnalytics, then set the idle time as infinite so idle state is not recorded.
-                Service.StartUp(product, new UserInfo(Session.UserId), hostInfo, TimeSpan.FromMinutes(30));
-                TrackPreferenceInternal("ReportingAnalytics", "", ReportingAnalytics ? 1 : 0);
+                Service.StartUp(product, new UserInfo(Session.UserId), hostInfo, TimeSpan.FromMinutes(30));                
             }
+
+            serviceInitialized.Set();
+        }
+        /// <summary>
+        /// Starts the client when DynamoModel is created. This method initializes
+        /// the Analytics service and application life cycle start is tracked.
+        /// </summary>
+        public void Start()
+        {
+            // Start Analytics service regardless of optin status.
+            // Each track event will be enabled/disabled based on the corresponding optin status.
+            // Ex. ADP will manage optin status internally
+            Task.Run(() => StartInternal());
+
+            TrackPreferenceInternal("ReportingAnalytics", "", ReportingAnalytics ? 1 : 0);
         }
 
         public void ShutDown()
         {
+            if (!Analytics.DisableAnalytics) serviceInitialized.Wait();
             Dispose();
         }
 
         public void TrackEvent(Actions action, Categories category, string description, int? value)
         {
-            if (!ReportingAnalytics) return;
+            if (Analytics.DisableAnalytics) return;
 
-            var e = AnalyticsEvent.Create(category.ToString(), action.ToString(), description, value);
-            e.Track();
+            Task.Run(() =>
+            {
+                serviceInitialized.Wait();
+
+                lock(trackEventLockObj)
+                {
+                    if (!ReportingAnalytics) return;
+
+                    var e = AnalyticsEvent.Create(category.ToString(), action.ToString(), description, value);
+                    e.Track();
+                }
+            });
         }
 
         public void TrackPreference(string name, string stringValue, int? metricValue)
         {
-            if (!ReportingAnalytics) return;
+            if (Analytics.DisableAnalytics) return;
+
+            Task.Run(() =>
+            {
+                serviceInitialized.Wait();
+
+                lock (trackEventLockObj)
+                {
+                    if (!ReportingAnalytics) return;
 
-            TrackPreferenceInternal(name, stringValue, metricValue);
+                    TrackPreferenceInternal(name, stringValue, metricValue);
```

Comment:
making these calls thread safe introduces the risk of deadlocks.
This one in particular looks a bit risky
`TrackPreferenceInternal` also uses the lock mechanism. I know Task.Run will execute the delegate code in the ThreadPool, so the calling code should probably finish before the delegate is executed on another thread.

Reply:
Ah, a bad merge, TrackPreferenceInternal should have been removed.